community/rfcs/25-03-14-GenAIExample-001-CodeTrans-with-Agents.md

```diff
@@ -0,0 +1,346 @@
+
+# prepare file paths for Java checkstyle
+CHECKSTYLE_JAR="./checkstyle.jar"
+CHECKSTYLE_CONFIG="./google_checks.xml"
+
+if [[ ! -f "$SOURCE_FILE" ]]; then
+  echo "Source file not found: $SOURCE_FILE"
+  exit 1
+fi
+
+case "$LANGUAGE" in
+  python)
+    echo "Running pylint..."
+    pylint "$SOURCE_FILE" > "$REPORT_FILE" 2>&1
+    ;;
+
+  javascript)
+    echo "Running eslint..."
+    eslint "$SOURCE_FILE" > "$REPORT_FILE" 2>&1
+    ;;
+
+  java)
+    echo "Running checkstyle..."
+    if [[ ! -f "$CHECKSTYLE_JAR" ]]; then
+      echo "Missing checkstyle.jar. Please download it first."
+      exit 1
+    fi
+    java -jar "$CHECKSTYLE_JAR" -c "$CHECKSTYLE_CONFIG" "$SOURCE_FILE" > "$REPORT_FILE" 2>&1
+    ;;
+
+  go)
+    echo "Running golangci-lint..."
+    golangci-lint run "$SOURCE_FILE" > "$REPORT_FILE" 2>&1
+    ;;
+
+  cpp)
+    echo "Running cpplint..."
+    cpplint "$SOURCE_FILE" > "$REPORT_FILE" 2>&1
+    ;;
+
+  *)
+    echo "Unsupported language: $LANGUAGE"
+    echo "Supported language: python, javascript, java, go, cpp"
+    exit 1
+    ;;
+esac
+
+echo "Lint check completed. Report saved to $REPORT_FILE"
+
+```
+
+#### Code Execution Tool
+
+Currently we only design to support code execution tool for `Python`.
+
+* Prevent code injection
+
+  * Use Python Abstract Syntax Tree (AST) to detect and block dangerous operations such as `import os`, `exec`, and `__import__`.
+
+    ```python
+    import ast
+
+    code = "xxx"
+    tree = ast.parse(code)
+    # analyze each node in ast tree
+    for node in ast.walk(tree):
+        # Do import checks
+    ```
+* Install dependencies automatically
+
+  * Code execution tool need to support extracting and installing dependencies from source code automatically.
+  * Use AST here to extract the `import`/`from xxx import xxx` libraries.
+
+    ```python
+    import ast
+
+    def extract_imports(code_str):
+        tree = ast.parse(code_str)
+        imports = set()
+        for node in ast.walk(tree):
+            if isinstance(node, ast.Import):
+                for alias in node.names:
+                    imports.add(alias.name.split('.')[0])
+            elif isinstance(node, ast.ImportFrom) and node.module:
+                imports.add(node.module.split('.')[0])
+        return list(imports)
+    ```
+* Sandbox execution
+
+  * To ensure that the code runs in a fully isolated environment, the tool needs to use container-based sandbox like `Docker`.
+  * In the context of resource constraints, the resource limits (including memory, CPU, process numbers) and security policies are needed.
+  * Since the dependencies need to be installed first, the network authority will be processed in two stages:
+    1. preperation: install the dependencies into a mounted path
+       ```bash
+       # add memory, cpu, process limit, and disable privilege operations
+       docker run --rm -m 512m --cpus="1" --pids-limit=128 \
+         --security-opt no-new-privileges -v $(pwd)/code:/code \
```

Comment:
This should use only the necessary capabilities, and drop all others:
`--cap-drop ALL --cap-add <what is needed>`